### PR TITLE
Added first version of settings v2

### PR DESCRIPTION
**Pull request by garethahealy**
https://github.com/DozerMapper/dozer/issues/463

### CODE REVIEW

**Comment by garethahealy** [comment]
First draft of giving the user control on how they want to set certain settings.

CC @orange-buffalo

File: core/src/main/java/org/dozer/config/resolvers/SystemEnvironmentSettingsResolver.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    private final SettingsResolver fallthrough;
+
+    public SystemEnvironmentSettingsResolver(SettingsResolver fallthrough) {
+        this.fallthrough = fallthrough;
+    }
+
+    @Override
+    public void init() {
+        //noop
```
Comment (orange-buffalo) [inline core/src/main/java/org/dozer/config/resolvers/SystemEnvironmentSettingsResolver.java]: In some implementations we do `fallthrough.init();`, in other do not. Maybe we should follow the same approach everywhere? 
Also in my understanding `fallthrough` is optional, so in places we do init it, we should probably check for `null`.

File: core/src/main/java/org/dozer/config/resolvers/YAMLSettingsResolver.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    private Map<String, Object> properties = new HashMap<>();
+
+    public YAMLSettingsResolver(SettingsResolver fallthrough, DozerClassLoader classLoader, String configFile) {
+        this.fallthrough = fallthrough;
+        this.classLoader = classLoader;
+        this.configFile = configFile;
+    }
+
+    @Override
+    public void init() {
```
Comment (orange-buffalo) [inline core/src/main/java/org/dozer/config/resolvers/YAMLSettingsResolver.java]: I would suggest to split this method by responsibilities, i.e. locating/reading the file and processing map from yaml.

File: core/src/main/java/org/dozer/config/processors/SettingsProcessor.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.dozer.config.processors;
+
+import org.dozer.config.Settings;
+
+public interface SettingsProcessor {
```
Comment (orange-buffalo) [inline core/src/main/java/org/dozer/config/processors/SettingsProcessor.java]: Would be great to have documentation for this interface.

File: core/src/main/java/org/dozer/config/processors/DefaultSettingsProcessor.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    private SettingsResolver getSettingsResolver() {
+        String configFileNameFound = System.getProperty(DozerConstants.CONFIG_FILE_SYS_PROP);
+        if (MappingUtils.isBlankOrNull(configFileNameFound)) {
+            configFileNameFound = DozerConstants.DEFAULT_CONFIG_FILE;
+        }
+
+        SettingsResolver defaultValueSettingsResolver = new DefaultValueSettingsResolver();
+        SettingsResolver systemEnvironmentSettingsResolver = new SystemEnvironmentSettingsResolver(defaultValueSettingsResolver);
```
Comment (orange-buffalo) [inline core/src/main/java/org/dozer/config/processors/DefaultSettingsProcessor.java]: Another alternative for implementing chained resolvers would be to have a "composite" resolver which iterates over all registered resolvers and returns the first non-null value. The benefit would be that we do not need to bother with `fallthrough` in every implementation.

**Comment by orange-buffalo** [comment]
@garethahealy, thanks a lot, looks very nice to me!

**Pull request merged**